Add support for Rust edition 2021.

m-ou-se opened the pull request:
This adds support for `ediiton = "2021"`.

Related `rustc` PR: https://github.com/rust-lang/rust/pull/79576

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @alexcrichton (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md) for more information.

alexcrichton commented:
Thanks for this! We'll probably want to hold off at least until the rustc PR is merged to avoid jumping the gun by accident, but one other question I have in the meantime is that we probably want this to be unstable until the edition is stabilized. Does rustc print an error if `--edition 2021` is used on the stable channel? If so we can probably just defer to rustc's stability and avoid having a separate stability check in Cargo.

m-ou-se commented:
> Does rustc print an error if --edition 2021 is used on the stable channel?

It should, yes!

Current stable rustc:
```
$ rustc --edition=2021
error: argument for `--edition` must be one of: 2015|2018. (instead was `2021`)
```

'Stable' rustc with https://github.com/rust-lang/rust/pull/79576:
```
$ rustc --edition=2021        
error: edition 2021 is unstable and only available for nightly builds of rustc.
```

alexcrichton commented:
Sounds great! r=me for when the corresponding rust-lang/rust PR is approved as well

m-ou-se commented:
Update: https://github.com/rust-lang/rust/pull/79576 is approved, but waiting for the beta week to end. It should be merged 2021-01-01.

m-ou-se commented:
https://github.com/rust-lang/rust/pull/79576 has been merged.

alexcrichton commented:
Great!

Could this also update this line?

https://github.com/rust-lang/cargo/blob/54631323dc2082710f1ab939cb887fdba5a6f363/src/cargo/ops/fix.rs#L699-L709

(or better yet refactor it so that the compiler warns us via enum-usage that the line needs to get updated)

Additionally do you know if lints are currently configured to enable this line yet -- https://github.com/rust-lang/cargo/blob/54631323dc2082710f1ab939cb887fdba5a6f363/src/cargo/ops/fix.rs#L648-L650 ?

m-ou-se commented:
Updated it to use the enum instead. Asking for an non-existent edition now gives an error instead of silently ignoring it like before.

I changed the `self.idioms && edition == "2018"` check to check for `>= 2018` for now. Some (all?) of these lints (`BARE_TRAIT_OBJECTS` etc.) might become denied in edition 2021. Depending on how that all works out, this condition in cargo might need to be changed back to `==` later.

m-ou-se commented on src/cargo/ops/fix.rs:
```diff
@@ -688,23 +688,19 @@ impl FixArgs {
+        match self.enabled_edition {
+            None | Some(Edition::Edition2015) => Edition::Edition2018,
+            Some(Edition::Edition2018) => Edition::Edition2021,
+            Some(Edition::Edition2021) => Edition::Edition2021,
```
This matches the old behaviour of *2018 → 2018*. Should this give an error instead?

m-ou-se commented:
The situation with `cargo fix` w.r.t. edition 2021 should get a bit more attention at some point. Now if you run it on a `edition = "2018"` crate, it'll tell you to switch back to 2015 so it can perform the edition fixes. But when 2021 is ready, it should be upgrading to 2021 instead, and can no longer warn the user for the mistake of placing `edition = "2018"` too soon in a *2015 → 2018* upgrade. Also, the error message for `2018 → 2021` with `edition = "2021"` should be different, as cargo will now tell the user to remote the `edition = ..` line instead of changing it back to `edition = "2018"`.

(Also, should `cargo fix` be able to go directly from 2015 to 2021, or only through 2018?)

alexcrichton commented:
Ok thanks for investigating! Seems like we should go ahead and land this and log the issues in a different location for handling later. Is there a central 2021 tracking issue we can log these bits on? Or should we create a new tracking issue in Cargo for these issues?

FWIW I'd imagine that we should probably stop printing an error and instead if you `fix --edition` when you're already in the latest edition it just says "success!" or something like that. For transitioning from  any edition to any future edition, I think it depends on the compiler lint support, but I'd imagine that we'd probably only want to transition one edition at a time to avoid any possible weird bugs.

I also think that we'll probably want `cargo fix --edition` in the 2018 edition to print a better error for now other than "unknown lint" or "needs stable rustc suport", we may want a bit in Cargo saying somewhere what the latest *stable* edition is so if you `fix --edition` in the latest stable edition then nothing happens (or Cargo just prints success or something like that)

m-ou-se commented:
> Is there a central 2021 tracking issue we can log these bits on?

Nope, but there's a [hackmd](https://hackmd.io/3eG6OZWHRbSMxoRxzwNhGQ?view) in which we're trying to collect everything around the 2021 edition. Eventually it should all be tracked in proper github issues collected on [the 2021 project board](https://github.com/orgs/rust-lang/projects/7), so let's just open an issue on this repository to keep track of this.

alexcrichton commented:
@bors: r+

👍 that looks perfect to me, thanks!

bors commented:
:pushpin: Commit bc0c820013860264fe6c0933daf0824649707423 has been approved by `alexcrichton`

<!-- @bors r=alexcrichton bc0c820013860264fe6c0933daf0824649707423 -->
<!-- homu: {"type":"Approved","sha":"bc0c820013860264fe6c0933daf0824649707423","approver":"alexcrichton"} -->

bors commented:
:hourglass: Testing commit bc0c820013860264fe6c0933daf0824649707423 with merge e427d01b4fab3b004b42446e607861516f67a92e...
<!-- homu: {"type":"BuildStarted","head_sha":"bc0c820013860264fe6c0933daf0824649707423","merge_sha":"e427d01b4fab3b004b42446e607861516f67a92e"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/cargo/runs/1652300738)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/cargo/runs/1652300738","builder_name":"checks-actions"} -->

alexcrichton commented:
@bors: retry

bors commented:
:hourglass: Testing commit bc0c820013860264fe6c0933daf0824649707423 with merge 3bb0968e2280510b7465fd05b9568053c74ba934...
<!-- homu: {"type":"BuildStarted","head_sha":"bc0c820013860264fe6c0933daf0824649707423","merge_sha":"3bb0968e2280510b7465fd05b9568053c74ba934"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/runs/1653115521)
Approved by: alexcrichton
Pushing 3bb0968e2280510b7465fd05b9568053c74ba934 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"alexcrichton","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/runs/1653115521"},"merge_sha":"3bb0968e2280510b7465fd05b9568053c74ba934"} -->
